<?xml version="1.0"?>
<robot name="morning_car" xmlns:xacro="http://ros.org/wiki/xacro">

<!-- Morning referens position, center of the rear wheels -->
<link name="base_link"/>

<!-- Hesai Pandar 64ch LiDAR -->
<joint name="base_roof_lidar" type="fixed">
    <parent link="base_link"/>
    <child link="hesai_lidar"/>
    <origin xyz="0.75 0 1.8" rpy="0 0 1.57"/>
</joint>

<link name="hesai_lidar">
    <visual>
      <geometry>
        <cylinder length="0.2" radius="0.2"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- FLIR Camera, Front Left -->
<joint name="base_flir_FL" type="fixed">
    <parent link="base_link"/>
    <child link="flir_front_left"/>
    <origin xyz="1.3 -0.5 1.27" rpy="1.57 0 0.785398"/>
</joint>

<link name="flir_front_left">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- FLIR Camera, Front Right -->
<joint name="base_flir_FR" type="fixed">
    <parent link="base_link"/>
    <child link="flir_front_right"/>
    <origin xyz="1.3 0.5 1.27" rpy="1.57 0 2.35619"/>
</joint>

<link name="flir_front_right">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- FLIR Camera, Rear Left -->
<joint name="base_flir_RL" type="fixed">
    <parent link="base_link"/>
    <child link="flir_rear_left"/>
    <origin xyz="0.15 -0.5 1.27" rpy="1.57 0 -0.785398"/>
</joint>

<link name="flir_rear_left">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- FLIR Camera, Rear Right -->
<joint name="base_flir_RR" type="fixed">
    <parent link="base_link"/>
    <child link="flir_rear_right"/>
    <origin xyz="0.15 0.5 1.27" rpy="1.57 0 -3.927"/>
</joint>

<link name="flir_rear_right">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- XSens MTi-100 imu -->
<joint name="base_imu" type="fixed">
    <parent link="base_link"/>
    <child link="imu_link"/>
    <origin xyz="0.35 0 1.25" rpy="0 0 0"/>
</joint>

<link name="imu_link">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- Active IR Camera -->
<joint name="base_active_ir_cam" type="fixed">
    <parent link="base_link"/>
    <child link="active_ir_camera_link"/>
    <origin xyz="2.0 0 0.99" rpy="1.57 0 1.57"/>
</joint>

<link name="active_ir_camera_link">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- AU RADAR, Front -->
<joint name="base_au_radar_front" type="fixed">
    <parent link="base_link"/>
    <child link="RADAR_FRONT"/>
    <origin xyz="3.05 0 0.44" rpy="0 0 -1.57"/>
</joint>

<link name="RADAR_FRONT">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- AU RADAR, Front Right -->
<joint name="base_au_radar_front_right" type="fixed">
    <parent link="base_link"/>
    <child link="RADAR_FRONT_RIGHT"/>
    <origin xyz="2.7 -0.85 0.44" rpy="0 0 2.61819"/>
</joint>

<link name="RADAR_FRONT_RIGHT">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- AU RADAR, Rear Right -->
<joint name="base_au_radar_rear_right" type="fixed">
    <parent link="base_link"/>
    <child link="RADAR_REAR_RIGHT"/>
    <origin xyz="-0.4 -0.85 0.44" rpy="0 0 2.356"/>
</joint>

<link name="RADAR_REAR_RIGHT">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- AU RADAR, Front Left -->
<joint name="base_au_radar_front_left" type="fixed">
    <parent link="base_link"/>
    <child link="RADAR_FRONT_LEFT"/>
    <origin xyz="2.7 0.73 0.44" rpy="0 0 0.5236"/>
</joint>

<link name="RADAR_FRONT_LEFT">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- AU RADAR, Rear Left -->
<joint name="base_au_radar_rear_left" type="fixed">
    <parent link="base_link"/>
    <child link="RADAR_REAR_LEFT"/>
    <origin xyz="-0.4 0.73 0.44" rpy="0 0 0.7854"/>
</joint>

<link name="RADAR_REAR_LEFT">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>

<!-- Luxonis Stereo Camera -->
<joint name="base_stereo_cam" type="fixed">
    <parent link="base_link"/>
    <child link="device_luxonis_oak_node_rgb_camera_optical_frame"/>
    <origin xyz="2.1 0 0.89" rpy="1.57 0 1.57"/>
</joint>

<link name="device_luxonis_oak_node_rgb_camera_optical_frame">
    <visual>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <material name="Red">
        <color rgba="1.0 0 0 1.0"/>
      </material>
    </visual>
</link>


</robot>
